File: src/main/scala/crosscut/MarkerGroupList.scala
package crosscut

import typings.obsidian.mod
import typings.obsidian.mod.FileSystemAdapter

import scala.scalajs.js.Dynamic.literal as l
import utils.Utils

import scala.collection.mutable
import concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.{Success, Failure}

/**
 * # object MarkerGroupList
 * Create a table of markers to the marker file.
 */
object MarkerGroupList:

  private type DOCNAME  = String
  private type DOCNAMES = mutable.HashSet[String]
  private type MARKER   = String

  /**
   * # MarkerGroupList
   * Create a table of marker and files that the markers appear in ^marker-00
   *
   * @param app
   * @param markerFile
   * @param documentFolder
   */
  def apply(app: mod.App, markerFile : String, documentFolder: String): Unit =
    val fsa = app.vault.adapter.asInstanceOf[FileSystemAdapter]
    val vaultPath = fsa.getBasePath()
    val markerFileNameWithPath = s"${markerFile}/marker-table.md"
    //
    // some containers to use later on
    //
    val markerToDocumentMap = mutable.HashMap[MARKER, DOCNAMES]()
    val allSolutionFiles = mutable.ArrayBuffer[MARKER]()
    //
    // get all the doc files to scan
    //
    val documentFiles = Utils.listMDFilesInVault(fsa, documentFolder)
    //
    // pick up all markers in the doc string doc file by doc file and aggregate the markers
    // before processing them
    //
    val allFutures = mutable.ListBuffer[Future[Unit]]()

    documentFiles.foreach(docFile =>

      allFutures += fsa.read(docFile).toFuture.map(fileContent =>

        val markersMatch = Utils.markerRegExp.findAllMatchIn(fileContent)
        val markersPerDocument = markersMatch.map(marker => fileContent.substring(marker.start, marker.end).trim).toList

        markersPerDocument.foreach(marker =>
          if !markerToDocumentMap.contains(marker) then
            markerToDocumentMap += (marker -> mutable.HashSet[String]())
          markerToDocumentMap(marker) += docFile
        )
      )
    )
    //
    // wait for all work to be done
    //
    val waitingForFutures : Future[List[Unit]] = Future.sequence(allFutures.toList)

    waitingForFutures.onComplete{
      case Success(_) =>
        val allMarkers = markerToDocumentMap
          .keys
          .toList
          .sortWith((s1, s2) =>
            s1 < s2
          )

        val mdString = StringBuilder(s"|marker|document|\n")
        mdString ++= s"|------|--------|\n"
        allMarkers.foreach(marker =>
          val docNameSet = markerToDocumentMap(marker)
          //
          // build marker to doc entry from the set of document names.
          //
          docNameSet.foreach(docName =>
            mdString ++= s"|${marker.drop(1)}|[[$docName#$marker]]\n"
          )
        )
        //
        // write of the solution text
        //
        Utils.makeDirInVault(fsa, markerFileNameWithPath)
        fsa.write(markerFileNameWithPath, mdString.toString()).toFuture.foreach(Unit => ())
      case Failure(ex) => println(s"Failed to complete all futures: ${ex.getMessage}")
    }

File: ts/MarkerGroupList.ts
import { App } from 'obsidian';
import { Utils } from './Utils'
import { DocFolders } from './DocFolders';

export class MarkerGroupList {

    markerFileWithPath : string;
    utils : Utils;
    docFolders : DocFolders;

    constructor(app : App, docFolders : DocFolders) {
        this.markerFileWithPath = `${docFolders.settingsMarkerFolder}/marker-table.md`
        this.utils = new Utils(app);
        this.docFolders = docFolders;
    }

    generateMakerGroupList() {
        //
        // some containers to use later on
        //
        const markerToDocumentMap = new Map<string, Set<string>>();
        const markerToTestDocumentMap = new Map<string, Set<string>>();
        //
        // get all the doc files to scan
        //
        const commentFiles = this.utils.listMDFilesInVault(this.docFolders.settingsCommentFolder);
        const testCommentFiles = this.utils.listMDFilesInVault(this.docFolders.settingsTestCommentFolder);
        //
        // pick up all markers in the doc string doc file by doc file and aggregate the markers
        // before processing them
        //
        const promise: Array<Promise<void>> = new Array<Promise<void>>();
        commentFiles.forEach(commentFile => {
            promise.push(this.utils.fsa.read(commentFile)
                .then(value => {
                        var markerSet = new Set<string>();
                        //
                        // get all the markers in the value string
                        //
                        const markersMatch = value.matchAll(this.utils.markerRegExp);
                        Array.from(markersMatch).forEach(marker => {
                            markerSet.add(marker[0].trim());
                        })
                        markerSet = this.utils.sortSetOfString(markerSet);
                        markerSet.forEach(marker => {
                            if (!markerToDocumentMap.has(marker.trim())) {
                                markerToDocumentMap.set(marker, new Set<string>());
                            }
                            markerToDocumentMap.get(marker.trim())?.add(commentFile);
                        });
                    }));
        })
        testCommentFiles.forEach(commentFile => {
            promise.push(this.utils.fsa.read(commentFile)
                .then(value => {
                        var markerSet = new Set<string>();
                        //
                        // get all the markers in the value string
                        //
                        const markersMatch = value.matchAll(this.utils.markerRegExp);
                        Array.from(markersMatch).forEach(marker => {
                            markerSet.add(marker[0].trim());
                        })
                        markerSet = this.utils.sortSetOfString(markerSet);
                        markerSet.forEach(marker => {
                            if (!markerToTestDocumentMap.has(marker.trim())) {
                                markerToTestDocumentMap.set(marker, new Set<string>());
                            }
                            markerToTestDocumentMap.get(marker.trim())?.add(commentFile);
                        });
                    }));
        })
        Promise.allSettled(promise)
            .then(value => {
                const allMarkers = Array.from(markerToDocumentMap.keys()).sort();

                var mdString = "Code and Code Story links\n\n";
                mdString = mdString + `|marker|document|\n`;
                mdString = mdString + `|------|--------|\n`;
                allMarkers.forEach(marker => {
                    const docNameSet = markerToDocumentMap.get(marker);
                    //
                    // build marker to doc entry from the set of document names.
                    //
                    docNameSet?.forEach(docName => {
                        mdString = mdString + `|${marker.substring(1)}|[[${docName}#${marker}]]\n`;
                    })
                })

                const allTestMarkers = Array.from(markerToTestDocumentMap.keys()).sort();

                mdString = mdString + "\nTest and Test Story links\n\n";
                mdString = mdString + `|marker|document|\n`;
                mdString = mdString + `|------|--------|\n`;
                allTestMarkers.forEach(marker => {
                    const docNameSet = markerToTestDocumentMap.get(marker);
                    //
                    // build marker to test doc entry from the set of document names.
                    //
                    docNameSet?.forEach(docName => {
                        mdString = mdString + `|${marker.substring(1)}|[[${docName}#${marker}]]\n`;
                    })
                })                
                this.utils.fsa.write(this.markerFileWithPath, mdString);
            })
    }
}